examples/cis2-dynamic-nft/src/lib.rs

```diff
@@ -0,0 +1,1450 @@
+        self.implementors.insert(std_id, implementors);
+    }
+}
+
+// Contract functions
+
+/// Initialize contract instance with a no token types.
+#[init(contract = "cis2_dynamic_nft", event = "Cis2Event<ContractTokenId, ContractTokenAmount>")]
+fn contract_init<S: HasStateApi>(
+    _ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+) -> InitResult<State<S>> {
+    // Construct the initial contract state.
+    Ok(State::empty(state_builder))
+}
+
+#[derive(Serialize, SchemaType)]
+struct ViewAddressState {
+    balances:  Vec<(ContractTokenId, ContractTokenAmount)>,
+    operators: Vec<Address>,
+}
+
+#[derive(Serialize, SchemaType)]
+struct ViewState {
+    state:  Vec<(Address, ViewAddressState)>,
+    tokens: Vec<ContractTokenId>,
+}
+
+/// View function for testing. This reports on the entire state of the contract
+/// for testing purposes. In a realistic example there `balance_of` and similar
+/// functions with a smaller response.
+#[receive(contract = "cis2_dynamic_nft", name = "view", return_value = "ViewState")]
+fn contract_view<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ReceiveResult<ViewState> {
+    let state = host.state();
+
+    let mut inner_state = Vec::new();
+    for (k, a_state) in state.state.iter() {
+        let mut balances = Vec::new();
+        let mut operators = Vec::new();
+        for (token_id, amount) in a_state.balances.iter() {
+            balances.push((*token_id, *amount));
+        }
+        for o in a_state.operators.iter() {
+            operators.push(*o);
+        }
+
+        inner_state.push((*k, ViewAddressState {
+            balances,
+            operators,
+        }));
+    }
+    let mut tokens = Vec::new();
+    for v in state.tokens.iter() {
+        tokens.push(*v.0);
+    }
+
+    Ok(ViewState {
+        state: inner_state,
+        tokens,
+    })
+}
+
+type TokenUpdateParams = ContractTokenId;
+
+///
+/// Only contract owner can call `upgrade` function to pint next item in the
+/// MetadataUrl list. It requires only tokenId, callable once at a time for a
+/// token.
+
+#[receive(
+    contract = "cis2_dynamic_nft",
+    name = "upgrade",
+    parameter = "TokenUpdateParams",
+    error = "ContractError",
+    enable_logger,
+    mutable
+)]
+fn contract_upgrade<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    // Get the contract owner
+    let owner = ctx.owner();
+    // Get the sender of the transaction
+    let sender = ctx.sender();
+
+    ensure!(sender.matches_account(&owner), ContractError::Unauthorized);
+
+    // Parse the parameter.
+
+    let token_id: TokenUpdateParams = ctx.parameter_cursor().get()?;
+
+    let state = host.state_mut();
+
+    // Mint the token in the state.
+    state.upgrade_token(&token_id).ok();
+
+    // Event for upgraded token.
```

Comment:
wrong comment, Not sure if we actually need the `TokenEvolvedEven`? Isn't `TokenMetadat` event emitting the same information.

Reply:
yes, removed the custom event. put that as an example, but TokenMeadata is providing the required functionality.